src/java/src/main/java/com/dnanexus/DXSearch.java

```diff
@@ -989,14 +1001,116 @@ private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
+                this.lastElementIndex = -1;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return lastElementIndex < (response.results.size() - 1);
+            }
+
+            @Override
+            public T next() {
+                return getDataObjectInstanceFromResult(response.results.get(++lastElementIndex));
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            @SuppressWarnings("unchecked")
+            private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
+                DXDataObject dataObject = null;
+                DXContainer container = DXContainer.getInstance(e.project);
+                if (e.describe != null) {
+                    dataObject =
+                            DXDataObject.getInstanceWithCachedDescribe(e.id, container, env,
+                                    e.describe);
+                } else {
+                    dataObject = DXDataObject.getInstanceWithEnvironment(e.id, container, env);
+                }
+
+                if (classConstraint != null) {
+                    if (!dataObject.getId().startsWith(classConstraint + "-")) {
+                        throw new IllegalStateException("Expected all results to be of type "
+                                + classConstraint + " but received an object with ID "
+                                + dataObject.getId());
+                    }
+                }
+                // This is an unchecked cast, but the callers of this class
+                // should have set T appropriately so that it agrees with
+                // the class constraint (if any). If something goes wrong
+                // here, either that code is incorrect or the API server
+                // has returned incorrect results.
+                return (T) dataObject;
+            }
+        }
+
+        /**
+         * Initializes result set page object with the specified page size.
+         */
+        private SearchPage(FindDataObjectsRequest request, String classConstraint,
+                                      DXEnvironment env, int pageSize) {
+            this.request = request;
+            this.classConstraint = classConstraint;
+            this.env = env;
+            this.pageSize = pageSize;
+            this.response = DXAPI.systemFindDataObjects(request, FindDataObjectsResponse.class, env);
+        }
+
+        /**
+         * Returns an amount of items on findDataObjects results page
+         */
+        int size() {
+            return response.results.size();
+        }
+
+        @Override
+        public Iterator<T> iterator()
+        {
+            return new ResultIterator();
+        }
+
+        public boolean hasNext() {
+            return response.next != null && !response.next.isNull();
+        }
+
+        public SearchPage<T> next() {
```

Comment:
I'd prefer if instead of having to serialize the whole SearchPage, the client could just get the same JSON thing that the API gives them, and then later resume the query by supplying the same (or equivalent) request and this deserialized value. e.g.

```
class SearchPage {
    public JsonNode getNext() { // Not sure what is a good name for this,
        // so as to not confuse it with the iterator "next"
}
```

And then above, in FindDataObjectsRequestBuilder, an additional method to resume the queries:

```
class FindDataObjectsRequestBuilder {
    public FindDataObjectsResult<T> execute() {...}

    public SearchPage<T> getFirstPage() {...}

    public SearchPage<T> getSubsequentPage(JsonNode starting) {...}
}
```